SIMSBIOHUB-567: Add/Support New Observation Environment Attributes

LouisThedroux opened the pull request:
## Links to Jira Tickets

https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-567

## Description of Changes

- Add new UI to support environment columns in observations table.
- Add new db tables for environment attribute codes.
- Inserted environmental variables and options as a new migration. 

## Testing Notes

Can add/remove environment attribute columns in observations table.
Can add/remove environment column cell values.
Can import observation csv with environment columns.

NickPhura committed:
Initial working observation import against sampling period

NickPhura committed:
Add better loading/disabled handlers

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-501

NickPhura committed:
Add Knip and SWC

NickPhura committed:
Update import observations

NickPhura committed:
Remove console logs

NickPhura committed:
Remove swc from API to fix unit tests

NickPhura committed:
Remove knip

NickPhura committed:
Remove gulp

NickPhura committed:
Update lock

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-501

mauberti-bc committed:
standards page and api service

NickPhura committed:
Merge branch 'dev' into SIMSBIOHUB-501

NickPhura committed:
Merge branch 'SIMSBIOHUB-501' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-501

NickPhura committed:
Add unit tests. fix spelling

NickPhura committed:
Add tests

mauberti-bc committed:
update standards page

mauberti-bc committed:
standards endpoint and service tests

mauberti-bc committed:
finish unit tests

mauberti-bc committed:
linter

mauberti-bc committed:
add standards to hamburger menu for smaller screens

mauberti-bc committed:
fix openapi spec

mauberti-bc committed:
linter

NickPhura committed:
Fix prettier (bump eslint/prettier versions)

NickPhura committed:
Update .eslintrc

NickPhura committed:
Update .eslintrc
Fix issues

NickPhura committed:
re-run install

NickPhura committed:
Merge branch 'dev' into npTechDebt

mauberti-bc committed:
Merge remote-tracking branch 'origin/npTechDebt' into taxon-standards

mauberti-bc committed:
prettier changes & node update

mauberti-bc committed:
Merge branch 'taxon-standards' of github.com:bcgov/biohubbc into taxon-standards

mauberti-bc committed:
linter

mauberti-bc committed:
Merge branch 'dev' of github.com:bcgov/biohubbc into taxon-standards

mauberti-bc committed:
attempt to solve prettier issues

mauberti-bc committed:
merge in dev

mauberti-bc committed:
basic configure table modal

mauberti-bc committed:
update file structure

mauberti-bc committed:
wip: measurement dialog

mauberti-bc committed:
merge in dev

mauberti-bc committed:
scroll behaviour

mauberti-bc committed:
refactor measurement search to be single select

mauberti-bc committed:
update width of observation table

mauberti-bc committed:
linter

mauberti-bc committed:
merge in dev

mauberti-bc committed:
fix merge issues

NickPhura committed:
WIP

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-567

NickPhura committed:
Working initial environment tables

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into configure-obs-table

NickPhura committed:
Merge branch 'configure-obs-table' into SIMSBIOHUB-567

NickPhura committed:
Add observation environment components, WIP.

NickPhura committed:
WIP wiring up backend

NickPhura committed:
WIP

NickPhura committed:
Mostly working

NickPhura committed:
Latest code

NickPhura committed:
Move code to hooks

mauberti-bc committed:
add environments lookup options

LouisThedroux committed:
init commit

NickPhura committed:
Update migration

NickPhura committed:
WIP

NickPhura committed:
Backend validation

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-567

NickPhura committed:
Fix existing tests

NickPhura committed:
Add unit tests

NickPhura committed:
ignore-skip

NickPhura committed:
Fix some code smells, remove unused code.

NickPhura committed:
Tweaks, Add unit tests, fix code smells

LouisThedroux committed:
added env vars

LouisThedroux committed:
Merge branch 'SIMSBIOHUB-567' of https://github.com/bcgov/biohubbc into env-variables-create

NickPhura committed:
Update observations table

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-567

LouisThedroux committed:
had to reinstate file losing ext. and being deleted. No idea what happened

LouisThedroux committed:
Merge branch 'SIMSBIOHUB-567' of https://github.com/bcgov/biohubbc into env-variables-create

NickPhura committed:
Update environment ids to uuids

NickPhura committed:
Update environment ids to uuids

NickPhura committed:
ignore-skip

LouisThedroux committed:
comma add

LouisThedroux committed:
Merge branch 'SIMSBIOHUB-567' of https://github.com/bcgov/biohubbc into env-variables-create

NickPhura committed:
Fix lint issue

LouisThedroux committed:
update env vars

LouisThedroux committed:
More env vars, styling chnages in measurements and standards cards.

LouisThedroux committed:
Merge branch 'SIMSBIOHUB-567' of https://github.com/bcgov/biohubbc into env-variables-create

LouisThedroux committed:
make fix

LouisThedroux committed:
merged in dev

LouisThedroux committed:
installs

LouisThedroux commented on database/src/migrations/20240502130000_env_variables_create.ts:
```diff
@@ -0,0 +1,2191 @@
+  'celsius'
+),
+(
+  'Sightability Correction Factor.',
```
Period

LouisThedroux committed:
formatting, editing syntax, merging migrations into 1

NickPhura committed:
Fix package-lock line endings

LouisThedroux committed:
remove duplicates

LouisThedroux committed:
Merge branch 'env-variables-create' of https://github.com/bcgov/biohubbc into env-variables-create

aubertinyoung commented on app/src/features/surveys/observations/observations-table/configure-columns/components/general/ConfigureGeneralColumns.tsx:
```diff
@@ -1,71 +1,74 @@
 import ListItemText from '@mui/material/ListItemText';
 import Stack from '@mui/material/Stack';
 import Typography from '@mui/material/Typography';
+import Box from '@mui/system/Box';
```
should be ```import Box from '@mui/material/Box';```

mauberti-bc committed:
remove last column in observations table

mauberti-bc committed:
styling for configure observation table dialog

mauberti-bc reviewed (dismissed):
Tested locally and worked as expected 👍

NickPhura committed:
Add jsdoc

NickPhura committed:
Add jsdoc 2

MacDeluca commented on app/src/features/surveys/observations/observations-table/configure-columns/components/measurements/useConfigureMeasurementColumns.tsx:
```diff
@@ -0,0 +1,103 @@
+          );
+
+          // Store all remaining measurement definitions in local storage
+          sessionStorage.setItem(
```
Once 572 goes in, you can use `usePersistentState` hook if you want

NickPhura replied:
Ill look at incorporating it in a separate PR, just to get his chunk in sooner.

I did start to drop it in, but in a few places I do some filtering of the state (which could maybe be removed, but not sure), that will require a bit more testing to make sure I don't accidentally break anything.

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into env-variables-create

NickPhura committed:
Merge conflicts

MacDeluca commented:
Code looks great
Few things I noticed while testing the app, you can omit them if they are intended or beyond the scope of this PR.

1. Nit: After refreshing the page the 'SHOW/HIDE ALL' button only works on second click
2. Hiding all the columns and refreshing the page doesnt keep the state (intended?)
3. Clicking configure and searching for 'antler configuration' -> then clicking Add Record crashes the UI with error: ```Cannot read properties of undefined (reading 'value')```. I think this is happening for both measurements and environment columns
4. Nit: The measurements in the dropdown could be capitalized similar to the environment options
5. Downloading CSV template and importing data throws error for this specific CSV. I very possibly could have formatted the data incorrectly.
[SIMS_Observations_Template_2024_05_22_10_28_07.csv](https://github.com/bcgov/biohubbc/files/15406897/SIMS_Observations_Template_2024_05_22_10_28_07.csv)
![image](https://github.com/bcgov/biohubbc/assets/99926243/94b44e6f-96fe-4067-b909-33ebf447da92)

NickPhura committed:
Code review fixes:
- Fix toggle visibility
- Fix visibility not persisting in session storage
- Fix csv import (default sheet logic)
- Fix capitalization nit

NickPhura commented:
@MacQSL 

Pushed up a commit with some fixes. I believe I've addressed everything you mentioned.

MacDeluca reviewed (approved):
LGTM 😺
